include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,23 +118,244 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
+
+                for (::std::uint16_t __idx = __item_id; __idx < __n; __idx += __wg_size)
+                {
+                   __out[__idx] = __lacc[__idx];
+                }
+
+                const ::std::uint16_t __residual = __n % __wg_size;
+                const ::std::uint16_t __residual_start = __n - __residual;
+                if (__residual > 0 && __item_id < __residual)
+                {
+                    auto __idx = __residual_start + __item_id;
+                   __out[__idx] = __lacc[__idx];
+                }
+            });
+        });
+        __event.wait();
+    }
+
+    static constexpr ::uint32_t __elems_per_wg = _ElemsPerItem * _WGSize;
+
+    template <typename... _Name>
+    using _GroupScanKernelBase = __par_backend_hetero::__scan_single_wg_kernel<__elems_per_wg, _IsFullGroup, _Inclusive, _Name...>;
+
+    template<typename _Policy, typename _InRng, typename _OutRng, typename _InitType, typename _BinaryOperation, typename _UnaryOp>
+    static void __launch_static_bounds_scan(const _Policy & __policy, _InRng&& __in, _OutRng&& __out, ::std::size_t __n, _InitType __init, _BinaryOperation __bin_op, _UnaryOp __unary_op)
+    {
+        using _ValueType = typename _InitType::__value_type;
+        using _CustomName = typename _Policy::kernel_name;
+        using _GroupScanKernel =
+                     oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<_GroupScanKernelBase, _CustomName, _BinaryOperation, _InRng, _OutRng>;
+
+        auto __event = __policy.queue().submit([&](sycl::handler& __hdl) {
+            oneapi::dpl::__ranges::__require_access(__hdl, __in, __out);
+
+            auto __lacc = __dpl_sycl::__local_accessor<_ValueType>(sycl::range<1>{__elems_per_wg}, __hdl);
+
+            __hdl.parallel_for<_GroupScanKernel>(sycl::nd_range<1>(_WGSize, _WGSize), [=](sycl::nd_item<1> __self_item) {
+                const auto& __group = __self_item.get_group();
+                const auto& __subgroup = __self_item.get_sub_group();
+                // This kernel is only launched for sizes less than 2^16
+                const ::std::uint16_t __item_id = __self_item.get_local_linear_id();
+                const ::std::uint16_t __subgroup_id = __subgroup.get_group_id();
+                const ::std::uint16_t __id_in_subgroup = __subgroup.get_local_id();
+                const ::std::uint16_t __subgroup_size = __subgroup.get_local_linear_range();
+
+                if constexpr (_IsFullGroup)
+                {
+                    _ONEDPL_PRAGMA_UNROLL
+                    for (::std::uint16_t __i = 0; __i < _ElemsPerItem; ++__i)
+                    {
+                       auto __idx = __i * _WGSize + __subgroup_id * __subgroup_size;
+                       auto __val = __unary_op(__subgroup.load(__in.begin() + __idx));
+                       __subgroup.store(__lacc.get_pointer() + __idx, __val);
```

Comment:
`syc::sub_group` class doesn't have yet such API - `store`.
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sub-group-class

The SYCL spec tell that such functionality there is for `vec` type only.
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:vector.type

So, we have to use `vec `type or wrap up that code into a kind of "Intel specific" macro.

Reply:
The subgroup load/store functions are different than the `vec` load/store instructions. I couldn't find a spec for an extension that would contain a macro for when this feature is available.

I will guard the use of subgroup load/store here on whether or not `__INTEL_COMPILER` is defined. Since this functionality is mentioned in [Intel's GPU optimization guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-gpu-optimization-guide/top/kernels/registers.html), it might be a safe assumption that if  `__INTEL_COMPILER` is defined then this feature is available.